### PR TITLE
[SPARK-33704][SQL] Support latest version of initialize() in HiveGenericUDTF

### PR DESCRIPTION
**Pull request by southernriver**
### What changes were proposed in this pull request?
For HiveGenericUDTF , there are two initialization methods:

```
  public StructObjectInspector initialize(StructObjectInspector argOIs)
      throws UDFArgumentException {
    List<? extends StructField> inputFields = argOIs.getAllStructFieldRefs();
    ObjectInspector[] udtfInputOIs = new ObjectInspector[inputFields.size()];
    for (int i = 0; i < inputFields.size(); i++) {
      udtfInputOIs[i] = inputFields.get(i).getFieldObjectInspector();
    }
    return initialize(udtfInputOIs);
  }
  @Deprecated
  public StructObjectInspector initialize(ObjectInspector[] argOIs)
      throws UDFArgumentException {
    throw new IllegalStateException("Should not be called directly");
  }
```
As https://issues.apache.org/jira/browse/HIVE-5737 mentioned, hive provided StructObjectInspector for UDTFs rather than ObjectInspect[], but Spark SQL  still only support deprecated function.

An exception will be reported before fix:

> Error in query: No handler for UDF/UDAF/UDTF 'FeatureParseUDTF1': java.lang.IllegalStateException: Should not be called directly
> Please make sure your function overrides public StructObjectInspector initialize(ObjectInspector[] args).; line 1 pos 7

This pr will resolve the exception.

### Why are the changes needed?

 For the migration of  Hive to Spark SQL，we face many UDTF functions throw exception because this issue，It’s really a great improvement for attracting users to Spark SQL.

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

manual

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @wangyum

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by wangyum** [comment]
@southernriver Could you make pr against master branch?

**Review by wangyum — CHANGES_REQUESTED** [review]
Remove `build/._scala-2.12.10` and `build/._zinc-0.3.15`.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 1]
```diff
@@ -91,8 +91,18 @@ private[sql] class HiveSessionCatalog(
...
+              udfExpr = Some(HiveGenericUDTF(name, new HiveFunctionWrapper(clazz.getName), input))
+              // Force it to check input data types.
+              udfExpr.get.asInstanceOf[HiveGenericUDTF].elementSchema
+            } catch {
+              case exception: Exception =>
+                logInfo(s"HiveGenericUDTF initialize(ObjectInspector[] args) is deprecated, and" +
+                  s" we will suit the latest version of initialize(StructObjectInspector argOIs).")
+                udfExpr = Some(HiveGenericUDTF(name, new HiveFunctionWrapper(clazz.getName),
+                  input, false))
+                udfExpr.get.asInstanceOf[HiveGenericUDTF].elementSchema
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:104]: How about?
```scala
val funcWrapper = new HiveFunctionWrapper(clazz.getName)
try {
  udfExpr = Some(HiveGenericUDTF(name, funcWrapper, input, true))
  // Force it to check data types.
  udfExpr.get.asInstanceOf[HiveGenericUDTF].elementSchema
} catch {
  case e: IllegalStateException if e.getMessage.equals("Should not be called directly") =>
    logInfo("Fallback to use the non deprecated UDTF constructor.")
    udfExpr = Some(HiveGenericUDTF(name, funcWrapper, input, false))
    // Force it to check data types.
    udfExpr.get.asInstanceOf[HiveGenericUDTF].elementSchema
}
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala [hunk 2]
```diff
@@ -198,7 +198,8 @@ private[hive] case class HiveGenericUDF(
 private[hive] case class HiveGenericUDTF(
     name: String,
     funcWrapper: HiveFunctionWrapper,
-    children: Seq[Expression])
+    children: Seq[Expression],
+    deprecated: Boolean = true)
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:202]: `deprecated: Boolean = true` -> `isDeprecatedConstructor: Boolean`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala [hunk 3]
```diff
@@ -212,7 +213,19 @@ private[hive] case class HiveGenericUDTF(
...
+  protected lazy val rowOI = ObjectInspectorFactory.getStandardStructObjectInspector(
+    inpuColNames, inputInspectors.asJava)
+
+  @transient
+  protected lazy val outputInspector =
+    if (deprecated) {
+      function.initialize(inputInspectors.toArray)
+    } else {
+      function.initialize(rowOI)
+    }
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala:228]: How about?
```scala
protected lazy val outputInspector = {
  if (isDeprecatedConstructor) {
    function.initialize(inputInspectors.toArray)
  } else {
    val rowOI = ObjectInspectorFactory.getStandardStructObjectInspector(
      children.zipWithIndex.map(e => s"_col${e._2}").asJava, inputInspectors.asJava)
    function.initialize(rowOI)
  }
}
```

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 4]
```diff
@@ -2132,11 +2132,9 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
            |USING JAR '${hiveContext.getHiveFile("SPARK-21101-1.0.jar").toURI}'
         """.stripMargin)
-      val e = intercept[org.apache.spark.sql.AnalysisException] {
-        sql("SELECT udtf_stack2(2, 'A', 10, date '2015-01-01', 'B', 20, date '2016-01-01')")
-      }
-      assert(
-        e.getMessage.contains("public StructObjectInspector initialize(ObjectInspector[] args)"))
+      val num =
+        sql("SELECT udtf_stack2(2, 'A', 10, date '2015-01-01', 'B', 20, date '2016-01-01')").count()
+      assert(num === 2)
```
Comment (wangyum) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:2137]: How about?
```scala
checkAnswer(
  sql("SELECT udtf_stack2(2, 'A', 10, date '2015-01-01', 'B', 20, date '2016-01-01')"),
  Seq(Row("A", 10, java.sql.Date.valueOf("2015-01-01")),
    Row("B", 20, java.sql.Date.valueOf("2016-01-01"))))
```

Comment (wangyum) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:2137]: Could you update test name?

**Comment by southernriver** [comment]
@wangyum  Thanks a lot,   that's  good catch. I'll  do some fix according to your suggestion.

**Comment by wangyum** [comment]
@southernriver Any update?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 5]
```diff
@@ -91,8 +91,18 @@ private[sql] class HiveSessionCatalog(
...
             udfExpr.get.dataType // Force it to check input data types.
           } else if (classOf[GenericUDTF].isAssignableFrom(clazz)) {
-            udfExpr = Some(HiveGenericUDTF(name, new HiveFunctionWrapper(clazz.getName), input))
-            udfExpr.get.asInstanceOf[HiveGenericUDTF].elementSchema // Force it to check data types.
+            try {
+              udfExpr = Some(HiveGenericUDTF(name, new HiveFunctionWrapper(clazz.getName), input))
+              // Force it to check input data types.
+              udfExpr.get.asInstanceOf[HiveGenericUDTF].elementSchema
+            } catch {
+              case exception: Exception =>
```
Comment (sunchao) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala:99]: why do we need this fallback mechanism? can we just switch to the new API without the `deprecated` flag?

**Review by sunchao — COMMENTED** [review]
I think this is trying to do the same thing as #29490

**Comment by wangyum** [comment]
@southernriver is busy these days. Let's close this.

**Pull request closed without merging**